### PR TITLE
Compile primitive

### PR DESCRIPTION
**Pull request by awni**
## Proposed changes

Compiled primitive with primitive-level fusion

- [x] Handle streams and max depth
- [x] Implement compile  vmap
- [x] Implement Compile vjp and jvp
- [x] Do the right thing for CPU compile (either skip compile entirely or fix the simulated tape to detach arrays and do donation properly)

### CODE REVIEW

**Comment by awni** [comment]
I think this is safe to review now. 
- Tests pass
- I'm happy with how CPU is handled (it still compiles but detaches etc)
- Transforms over Compiled work and for the sake of keeping the diff manageable, I would prefer to do that it in a follow on if everyone is 👌 .

File: mlx/compile.cpp [hunk 1]
```diff
@@ -327,6 +467,216 @@ void compile_simplify(
...
+    if (global_cache.find(arr.id()) != global_cache.end()) {
+      continue;
+    }
+
+    // Two pass recursion:
+    // First pass:
+    //  - Collect all the primitives which we can fuse with
+    //  - Keeps a cache of fusable primitives which may be added out of
+    //    DAG order. We have to determine if all of a fused primitive's
+    //    outptus are also in the fused section, and this may not be the
```
Comment (angeloskath) [inline mlx/compile.cpp:505]: nit: typo outptus -> outputs

File: mlx/compile.cpp [hunk 2]
```diff
@@ -327,6 +467,216 @@ void compile_simplify(
   }
 }
 
+// Extract sub-graphs of the graph that can be compiled
+// and replace them with a Compiled Primitive.
+void compile_fuse(
```
Comment (angeloskath) [inline mlx/compile.cpp:472]: Very very nice!

**Review by angeloskath — APPROVED** [review]
This is perfect! Also very well tested, I'm super hyped!

File: mlx/compile.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Copyright © 2023-2024 Apple Inc.
+
+#pragma once
+
+#include "mlx/array.h"
+
+namespace mlx::core {
+
+enum class CompileMode { disabled, no_simplify, no_fuse, enabled };
```
[the remark this answers is not in the document]
Comment (awni) [reply mlx/compile.h:9]: BTW I didn't expose this in Python yet.. but we can (and just not advertise it) if it's helpful for debugging. Mostly added it so I could test simplify and fusion separately.

**Pull request merged**